compat/win32/path-utils.c

```diff
@@ -50,3 +53,191 @@ int win32_offset_1st_component(const char *path)
+		goto fail;
+
+	if (read_at(fd, buffer, 0x3c, 4) < 0)
+		goto fail;
+
+	/* Read the `PE\0\0` signature and the COFF file header */
+	offset = le32(buffer);
+	if (read_at(fd, buffer, offset, 24) < 0)
+		goto fail;
+
+	if (buffer[0] != 'P' || buffer[1] != 'E' || buffer[2] != '\0' || buffer[3] != '\0')
+		goto fail;
+
+	num_sections = le16(buffer + 6);
+	opt_header_size = le16(buffer + 20);
+	offset += 24; /* skip file header */
+
+	/*
+	 * Validate magic number 0x10b or 0x20b, for full details see
+	 * https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#optional-header-standard-fields-image-only
+	 */
+	if (read_at(fd, buffer, offset, 2) < 0 ||
+	    ((i = le16(buffer)) != 0x10b && i != 0x20b))
+		goto fail;
+
+	offset += opt_header_size;
+
+	for (i = 0; i < num_sections; i++) {
+		if (read_at(fd, buffer, offset + i * 40, 40) < 0)
+			goto fail;
+
+		/*
+		 * For full details about the section headers, see
+		 * https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#section-table-section-headers
+		 */
+		if ((le32(buffer + 36) /* characteristics */ & ~0x600000) /* IMAGE_SCN_ALIGN_32BYTES */ ==
+		    (/* IMAGE_SCN_CNT_INITIALIZED_DATA */ 0x00000040 |
+		     /* IMAGE_SCN_MEM_READ */ 0x40000000 |
+		     /* IMAGE_SCN_MEM_WRITE */ 0x80000000)) {
+			size_t size = le32(buffer + 16); /* "SizeOfRawData " */
+			size_t pointer = le32(buffer + 20); /* "PointerToRawData " */
+
+			/*
+			 * Skip the section if either size or pointer is 0, see
+			 * https://github.com/golang/go/blob/go1.21.0/src/debug/buildinfo/buildinfo.go#L333
+			 * for full details.
+			 *
+			 * The size actually must be at least `buildInfoSize`,
+			 * i.e. 32.
+			 */
+			if (size < 32 || !pointer)
+				continue;
+
+			p = malloc(size);
+
+			if (!p || read_at(fd, p, pointer, size) < 0)
+				goto fail;
+
+			/*
+			 * Look for the build information embedded by Go, see
+			 * https://github.com/golang/go/blob/go1.21.0/src/debug/buildinfo/buildinfo.go#L165-L175
+			 * for full details.
+			 *
+			 * Note: Go contains code to enforce alignment along a
+			 * 16-byte boundary. In practice, no `.exe` has been
+			 * observed that required any adjustment, therefore
+			 * this here code skips that logic for simplicity.
+			 */
+			q = memmem(p, size - 18, "\xff Go buildinf:", 14);
+			if (!q)
+				goto fail;
+			/*
+			 * Decode the build blob. For full details, see
+			 * https://github.com/golang/go/blob/go1.21.0/src/debug/buildinfo/buildinfo.go#L177-L191
+			 *
+			 * Note: The `endianness` values observed in practice
+			 * were always 2, therefore the complex logic to handle
+			 * any other value is skipped for simplicty.
+			 */
+			if ((q[14] == 8 || q[14] == 4) && q[15] == 2) {
+				/*
+				 * Only handle a Go version string with fewer
+				 * than 128 characters, so the Go UVarint at
+				 * q[32] that indicates the string's length must
+				 * be only one byte (without the high bit set).
+				 */
+				if ((q[32] & 0x80) ||
```

Comment:
I think there may still be a chance this `q[32]` would read past the end of `p`.

There's now a check above to make sure `size` is at least 32, but that doesn't guarantee more than 32 bytes, and the `UVarint` which prefixes the Go version string starts _after_ the 32-byte thing I've been calling the "Go header".

Consider the case where `size` is 32 exactly, so we know `p` includes the 32-byte Go header.  The `memmem()` search is now restricted from looking at the last 18 bytes of `p`, so if there's a match and `size == 32`, it must be on the first 14 bytes of `p`, which means `q == p`.  So then `q[32]` is the same as `p[32]`, but since we're considering a case where `size == 32`, that's beyond the end of `p`.

In my suggested code in my [previous](https://github.com/git-for-windows/git/pull/5042#discussion_r1668151705) [comments](https://github.com/git-for-windows/git/pull/5042#discussion_r1669277035), I had three checks: first for the 32 bytes of the Go header, then for a single following byte (the start of the `UVarint`), and then for the length of the subsequent string.  My indexing was different, and I'd decreased the value of `size` to make the logic a bit more apparent, so after validating the `q[14]` and `q[15]` values, I had this before reading the equivalent of `q[32]`:

```c
				if (size < 1)
					goto fail;
```

So, either maybe do something like:
```suggestion
				if ((q + 33 - p) > size ||
				    (q[32] & 0x80) ||
```

Or, change the `if (size < 32 || !pointer)` above to `if (size < 33 || !pointer)`.  In that case, I'd add a note to the preceding comment to explain that we're ensuring both a 32-byte Go header and at least the first byte of a  `UVarint` that's expected to follow.

Reply:
That's a good point. I'll change that `size < 32` condition, but not only to require 33 bytes (that would only leave space for the UVarInt but not for any non-empty string), but require enough space for `go1`.